crates/gitbutler-repo/src/hooks.rs

```diff
@@ -0,0 +1,86 @@
+use std::path::PathBuf;
+
+use anyhow::Result;
+use git2_hooks;
+use git2_hooks::HookResult as H;
+use gitbutler_command_context::CommandContext;
+use gitbutler_diff::GitHunk;
+use serde::Serialize;
+
+use crate::staging;
+
+#[derive(Serialize, PartialEq, Debug, Clone)]
+pub struct MessageData {
+    pub message: String,
+}
+
+#[derive(Serialize, PartialEq, Debug, Clone)]
+pub struct ErrorData {
+    pub error: String,
+}
+
+/// Hook result indicating either success or failure.
+#[derive(Serialize, PartialEq, Debug, Clone)]
+#[serde(tag = "status", rename_all = "lowercase")]
+pub enum HookResult {
+    Success,
+    NotConfigured,
+    Message(MessageData),
+    Failure(ErrorData),
+}
+
+pub fn message(ctx: &CommandContext, mut message: String) -> Result<HookResult> {
+    let original_message = message.clone();
+    match git2_hooks::hooks_commit_msg(ctx.repo(), Some(&["../.husky"]), &mut message)? {
+        H::Ok { hook: _ } => match message == original_message {
+            true => Ok(HookResult::Success),
+            false => Ok(HookResult::Message(MessageData { message })),
+        },
+        H::NoHookFound => Ok(HookResult::NotConfigured),
+        H::RunNotSuccessful { stdout, stderr, .. } => {
+            let error = join_output(stdout, stderr);
+            Ok(HookResult::Failure(ErrorData { error }))
+        }
+    }
+}
+
+pub fn pre_commit(
+    ctx: &CommandContext,
+    selected_hunks: &Vec<(PathBuf, Vec<GitHunk>)>,
+) -> Result<HookResult> {
+    let mut index = ctx.repo().index()?;
+    staging::stage(ctx, selected_hunks)?;
+    let result = match git2_hooks::hooks_pre_commit(ctx.repo(), Some(&["../.husky"]))? {
+        H::Ok { hook: _ } => Ok(HookResult::Success),
+        H::NoHookFound => Ok(HookResult::NotConfigured),
+        H::RunNotSuccessful { stdout, stderr, .. } => {
+            let error = join_output(stdout, stderr);
+            Ok(HookResult::Failure(ErrorData { error }))
+        }
+    };
+    // Reset to original index to unstage files.
+    index.write()?;
```

Comment:
If the code in `git2_hooks::hooks_pre_commit` panics, this function will not have the opportunity to invoke  `index.write()` and reset the index.
To combat this, it seems like you can implement the [Drop](https://doc.rust-lang.org/book/ch15-03-drop.html) trait on something.

I can think of 3 ways to achieve this here:
1. Create a new struct `Guard` for which you implement drop and have it constructed at the beginning of the call and then let it go out of scope at the end.
2. Implement `pre_commit` as a method on a Struct that implements drop
3. Perhaps use something like the crate [scopeguard](https://docs.rs/scopeguard/latest/scopeguard/) 

I have personally not done any of the 3 things outlined. Maybe @Byron has some good advice for how to approach this

Reply:
I've done this using scopeguard now, thanks.